salt/utils/vmware.py

```diff
@@ -158,104 +166,217 @@ def esxcli(host, user, pwd, cmd, protocol=None, port=None, esxi_host=None):
+                         domain=None):
+    '''
+    Authenticate with a vCenter server or ESX/ESXi host and return the service instance object.
+
+    host
+        The location of the vCenter server or ESX/ESXi host.
+
+    username
+        The username used to login to the vCenter server or ESX/ESXi host.
+        Required if mechanism is ``userpass``
+    password
+        The password used to login to the vCenter server or ESX/ESXi host. This
+        Required if mechanism is ``sspi``
+
+    protocol
+        Optionally set to alternate protocol if the vCenter server or ESX/ESXi host is not
+        using the default protocol. Default protocol is ``https``.
+
+    port
+        Optionally set to alternate port if the vCenter server or ESX/ESXi host is not
+        using the default port. Default port is ``443``.
+
+    mechanism
+        pyVmomi connection mechanism. Can either be ``userpass`` or ``sspi``
+
+    principal
+        Kerberos service principal. Required if mechanism is ``sspi``
+
+    domain
+        Kerberos user domain. Required if mechanism is ``sspi``
+    '''
+
+    if protocol is None:
+        protocol = 'https'
+    if port is None:
+        port = 443
+
+    service_instance = GetSi()
+    if service_instance:
+        if service_instance._GetStub().host == ':'.join([host, str(port)]):
+            log.trace('Retrieving cached service instance')
+            service_instance._GetStub().GetConnection()
+        else:
+            Disconnect(service_instance)
+            service_instance = None
+
+    if not service_instance:
+        service_instance = _get_service_instance(host,
+                                                 username,
+                                                 password,
+                                                 protocol,
+                                                 port,
+                                                 mechanism,
+                                                 principal,
+                                                 domain)
+
+    # Test if data can actually be retrieved or connection has gone stale
+    log.trace('Checking connection is still authenticated')
+    try:
+        service_instance.CurrentTime()
+    except vim.fault.NotAuthenticated:
+        log.trace('Session no longer authenticating. Reconnecting')
+        Disconnect(service_instance)
+        service_instance = _get_service_instance(host,
+                                                 username,
+                                                 password,
+                                                 protocol,
+                                                 port,
+                                                 mechanism,
+                                                 principal,
+                                                 domain)
+
+    return service_instance
+
+
+def disconnect(service_instance):
```

Comment:
No need for this function. ``Disconnect`` is already imported into the namespace.